pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/LeaderElectionImpl.java

```diff
@@ -111,13 +112,13 @@ private synchronized CompletableFuture<LeaderElectionState> elect() {
             } else {
                 return tryToBecomeLeader();
             }
-        }).thenComposeAsync(leaderElectionState -> {
+        }).thenCompose(leaderElectionState -> {
```

Comment:
Using thenCompose makes it basically unpredictable the thread that executed the callback. 
This is way CompletableFuture sucks about threading.
If the Future is already completed then the callback is executed immediately.
Sometimes the callback is executed in the thread that completes the future, and something it is executed in any thread that calls 'get'

This fix may work.
I don't have time today to take a more careful look

Reply:
Hi, @eolivelli 
I think it is predictable, just more than one case.
The rule of the `CompletableFuture` is it will use the previous thread to do the next work because it is just simply run the next task in the `executor` or `current thread`. So, there are two cases here:

1. Run with the `metadata` thread. 
2. Run with the `get` thread.

In case 1, there is no problem because the `get` thread is blocking on get operation and waiting for the metadata thread to complete the rest of the work to notify the future.

In case 2, as you mentioned, if some future is already complete, it will still use the `get` thread to do the rest of the work. It also means the `CompletetableFuture` fallback to the synchronous operation. And the `get` operation is always behind the rest of the work. 

Honestly, we should make that behaviour asynchronous to avoid the deadlock. But we should change a lot of logic to ensure the notification order. (I think I will do it later)

I wonder if I made it clear. Please let me know if I missed something.